app/webhook_test.go

```diff
@@ -470,3 +473,164 @@ func TestSplitWebhookPost(t *testing.T) {
+		Description:  "some-description",
+		CreatorId:    th.BasicUser.Id,
+	}
+
+	th.App.UpdateConfig(func(cfg *model.Config) { cfg.ServiceSettings.EnableOutgoingWebhooks = true })
+
+	createdHook, err := th.App.CreateOutgoingWebhook(&outgoingWebhook)
+
+	if err != nil {
+		t.Fatalf("should not have failed: %v", err.Error())
+	}
+
+	assert.NotNil(t, createdHook, "should not be null")
+
+	assert.Equal(t, createdHook.ChannelId, outgoingWebhook.ChannelId)
+	assert.Equal(t, createdHook.TeamId, outgoingWebhook.TeamId)
+	assert.Equal(t, createdHook.CallbackURLs, outgoingWebhook.CallbackURLs)
+	assert.Equal(t, createdHook.Username, outgoingWebhook.Username)
+	assert.Equal(t, createdHook.IconURL, outgoingWebhook.IconURL)
+	assert.Equal(t, createdHook.DisplayName, outgoingWebhook.DisplayName)
+	assert.Equal(t, createdHook.Description, outgoingWebhook.Description)
+
+}
+
+func TestTriggerOutGoingWebhookWithUsernameAndIconURL(t *testing.T) {
+	th := Setup().InitBasic()
+	defer th.TearDown()
+
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.ServiceSettings.AllowedUntrustedInternalConnections = "localhost 127.0.0.1"
+	})
+	createdPost := make(chan *model.Post)
+
+	for name, testCase := range getTestCases() {
+		t.Run(name, func(t *testing.T) {
+
+			th.App.UpdateConfig(func(cfg *model.Config) {
+				cfg.ServiceSettings.EnableOutgoingWebhooks = true
+				cfg.ServiceSettings.EnablePostUsernameOverride = testCase.EnablePostUsernameOverride
+				cfg.ServiceSettings.EnablePostIconOverride = testCase.EnablePostIconOverride
+			})
+
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Write([]byte(`{"text": "sample response text from test server"}`))
+			}))
+			defer ts.Close()
+
+			channel := th.CreateChannel(th.BasicTeam)
+			hook, _ := createOutgoingWebhook(channel, ts.URL, th)
+			payload := getPayload(hook, th, channel)
+
+			th.App.TriggerWebhook(payload, hook, th.BasicPost, channel)
+
+			waitUntilWebhookResposeIsCreatedAsPost(channel, th, t, createdPost)
+
+			select {
+			case webhookPost := <-createdPost:
+				assert.Equal(t, webhookPost.Message, "sample response text from test server")
+				assert.Equal(t, webhookPost.Props["from_webhook"], "true")
+				if testCase.ExpectedIconUrl != "" {
+					assert.Equal(t, webhookPost.Props["override_icon_url"], testCase.ExpectedIconUrl)
+				} else {
+					assert.Nil(t, webhookPost.Props["override_icon_url"])
+				}
+
+				if testCase.ExpectedUsername != "" {
+					assert.Equal(t, webhookPost.Props["override_username"], testCase.ExpectedUsername)
+				} else {
+					assert.Nil(t, webhookPost.Props["override_username"])
+				}
+			case <-time.After(5 * time.Second):
+				t.Fatal("Timeout, webhook response not created as post")
+			}
+
+		})
+	}
+}
+
+func waitUntilWebhookResposeIsCreatedAsPost(channel *model.Channel, th *TestHelper, t *testing.T, createdPost chan *model.Post) {
+	go func() {
+		for i := 0; i < 5; i++ {
+			time.Sleep(time.Second)
```

Comment:
From experience, this kind of test will eventually break the build. I wonder if we might be able to leverage the new plugin capability to intercept a `MessageHasBeenPosted` hook within the tests? 0/5 as to whether this is part of this PR.

Reply:
so.. Should we handle it as a separate PR ?

Reply:
Sounds good.